lib/two_factor_authentication/models/two_factor_authenticatable.rb

```diff
@@ -11,17 +11,37 @@ module ClassMethods
         def has_one_time_password(options = {})
           include InstanceMethodsOnActivation
           include EncryptionInstanceMethods if options[:encrypted] == true
-
-          before_create { populate_otp_column }
         end
 
         ::Devise::Models.config(
-          self, :max_login_attempts, :allowed_otp_drift_seconds, :otp_length,
-          :remember_otp_session_for_seconds, :otp_secret_encryption_key)
+          self, :max_login_attempts, :allowed_otp_drift_seconds, :otp_length, :direct_otp_length,
+          :direct_otp_valid_for, :remember_otp_session_for_seconds, :otp_secret_encryption_key)
       end
 
       module InstanceMethodsOnActivation
         def authenticate_otp(code, options = {})
+          return true if direct_otp and authenticate_direct_otp(code, options)
+          return true if totp_enabled? and authenticate_totp(code, options)
+          false
+        end
+
+        def authenticate_direct_otp(code, options = {})
+          return false if direct_otp.nil?
+
+          return false if direct_otp != code
```

Comment:
Can these be combined, like this? `return false if direct_otp.nil? || direct_otp != code`